test-tests/shared/src/test/scala/zio/test/laws/LawsFSpec.scala

```diff
@@ -0,0 +1,74 @@
+package zio.test.laws
+
+import zio.ZIO
+import zio.random.Random
+import zio.test.Assertion._
+import zio.test._
+
+object LawsFSpec extends ZIOBaseSpec {
+
+  val OptionGenF: GenF[Random, Option] =
+    new GenF[Random, Option] {
+      def apply[R1 <: Random, A](gen: Gen[R1, A]): Gen[R1, Option[A]] =
+        Gen.option(gen)
+    }
+
+  trait Covariant[F[+_]] {
+    def map[A, B](f: A => B): F[A] => F[B]
+  }
+
+  object Covariant extends LawfulF.Covariant[Covariant] {
+
+    val identityLaw = new LawsF.Covariant.Law1[Covariant]("identityLaw") {
+      def apply[F[+_]: Covariant, A](fa: F[A]): TestResult = {
```

Comment:
In general in order to perform these checks, we would require type class instances for the parametric `A`, such as `Equal`, for example. This to me suggests maybe `Law1` etc should be parameterized by two things:

The higher-kinded type class (in this example, `Covariant`), and the set of type classes that must be satisfied by the parametric types (e.g. `Equal`).

Although that does raise other complications, such as where do these types come from. If the user has control of the type classes, then they also need to have control over the "poly" gens for the types.

Reply:
Yes I was thinking about that as well. I think the way we solve it is with `EqualF`, which gives us a way to determine whether two `F[A]` values are equal given a way to determine whether two `A` values are equal. I don't think it is enough just to have a way to determine whether two `A` values are equal because we don't know how to compare the `F[A]` values. We could implement this similarly to how we did `GenF`.

Reply:
I do think we need `EqualF` and friends.. then we can compare the `F[A]` given an `Equal: A`. But where will we get that `Equal[A]`? The function is polymorphic in `A` so has no structure on A`. So it seems we need them both: propagating capabilities on the types down, but also polymorphic basic type classes like `EqualF`.

Reply:
Yes. Ideally from the perspective of the caller we would like to remain fully polymorphic in the `A` type. Potentially this aspect of the functionality could best be provided along with the library that defines the type classes. Then we could say "the law has to operate on abstract types `A`, `B`, and `C` and the only guarantee we will give you is they have this structure". And then under the hood the test framework just uses something like integers that have all this structure.

Reply:
I think you are right. The downside of that is I think that means we will need the caller to provide us with a generator of `A` values with the required capabilities. We will then take it and hide the actual type in everything we do but we need some way to generate values with the required capabilities.

Reply:
Yes, agreed. And it's unfortunate, but it kind of makes sense, if you think about constrained functors, for example, you can only map them to specify types (e.g. types that are serializable), so the user is going to have to ultimately provide the gens for values that have the required properties, but like you say, we can hide it in the actual laws.

I briefly thought about having something like `G <: GenPoly`. If you propagated that in sufficient places, you could put the type class instances in there. Maybe not quite as nice to use from a syntactic perspective, because the instances aren't implicit, but maybe better integrated with the `GenPoly` machinery.

Reply:
Yes that is definitely something we could explore.